src/cmd_vel_mux.cpp

```diff
@@ -36,31 +36,56 @@ namespace cmd_vel_mux
 
 CmdVelMux::CmdVelMux(rclcpp::NodeOptions options) : rclcpp::Node("cmd_vel_mux", options.allow_undeclared_parameters(true).automatically_declare_parameters_from_overrides(true)), allowed_(VACANT)
 {
-  std::vector<std::string> names;
-  if (!get_parameter("subscribers.name", names))
+  std::map<std::string, rclcpp::Parameter> subs_parameters;
+  std::map<std::string, std::string> topics;
+  std::map<std::string, double> timeouts;
+  std::map<std::string, int64_t> priorities;
+  std::map<std::string, std::string> short_descs;
+
+  if (!get_parameters("", subs_parameters) || subs_parameters.size() <= 1)
   {
     RCLCPP_WARN(get_logger(), "No subscribers configured!");
   }
-
-  std::vector<std::string> topics;
-  get_parameter("subscribers.topic", topics);
-
-  std::vector<double> timeouts;
-  get_parameter("subscribers.timeout", timeouts);
-
-  std::vector<int64_t> priorities;
-  get_parameter("subscribers.priority", priorities);
-
-  std::vector<std::string> short_descs;
-  get_parameter("subscribers.short_desc", short_descs);
-
-  if (names.size() != topics.size() || names.size() != timeouts.size() ||
-      names.size() != priorities.size() || names.size() != short_descs.size())
+  else
   {
-    throw std::runtime_error("Number of names, topics, timeouts, priorities, and short_descs must be the same");
+    for(auto& x : subs_parameters)
+    {
+      std::string name = x.first;
+      rclcpp::Parameter parameter = x.second;
+      if (name.find(".topic") != std::string::npos && parameter.get_type() == rclcpp::ParameterType::PARAMETER_STRING)
+      {
+        std::string vector_id = name.substr(12, name.find(".topic")-12);
+        if (!get_parameter("subscribers." + vector_id + ".timeout", timeouts[vector_id]))
+        {
+          RCLCPP_WARN(get_logger(), "Subscriber " + vector_id + " has no timeout, not adding");
+          continue;
+        }
+        if (!get_parameter("subscribers." + vector_id + ".priority", priorities[vector_id]))
+        {
+          RCLCPP_WARN(get_logger(), "Subscriber " + vector_id + " has no priority, not adding");
+          continue;
+        }
+        bool priority_dup_flag = false;
+        for (auto it = priorities.begin(); it != priorities.end(); ++it)
+        {
+          if (it->second == priorities[vector_id] && it != priorities.find(vector_id)) priority_dup_flag = true;
+        }
+        if (priority_dup_flag)
+        {
+          RCLCPP_WARN(get_logger(), "Subscriber " + vector_id + " has the same priority as another subscriber, not adding");
+          continue;
+        }
+        if (!get_parameter("subscribers." + vector_id + ".short_desc", short_descs[vector_id]))
+        {
+          short_descs[vector_id] =  "No description";
+        }
+        topics[vector_id] = parameter.as_string();
+        parameters_names.push_back(vector_id);
+      }
+    }
```

Comment:
This loop is pretty inefficient.  It's not hugely important from a performance point-of-view, but I think it would be easier to understand if the loop did something with every single entry, instead of only doing work for the "topic" one and then using `get_parameter` to get the rest.

I'm suggesting something more like the following pseudo-code:

```
for (p in parameters)
{
    split name into 3 parts: subscribers. , "input name". , "variable"
    if (input_name does not exist) {
        list_.emplace_back(new CmdVelSub(input_name));
        index = list_.size() - 1;
    } else {
       index = list_.find(input_name);
    }
    list_[index].variable_name = variable.value;
}

configureFromList();  // Note that this no longer takes any parameters, as it just walks the list with the filled-in values and sets up subscribers and timers appropriately.
```

Let me know if this makes sense; it's a large restructuring of the code, but I think it makes more sense.

Reply:
+10